src/LibUsbSharp/Internal/RundownGuard.cs

```diff
@@ -0,0 +1,441 @@
+    ///     {
+    ///         // This thread performs the teardown exclusively
+    ///         // ... dispose resources, stop I/O, etc.
+    ///     }
+    ///     // If rd is null, another thread is handling teardown; this call blocks until completion.
+    /// }
+    /// ]]></code>
+    /// </example>
+    public IDisposable? WaitForRundown()
+    {
+        lock (_lock)
+        {
+            // If already completed, nothing to do.
+            if (_rundownCompleted)
+                return null;
+
+            // If another thread is handling rundown, wait until it completes, then return null.
+            if (_rundownStarted)
+            {
+                while (!_rundownCompleted)
+                {
+                    _ = Monitor.Wait(_lock);
+                }
+                return null;
+            }
+
+            // Become the rundown owner.
+            _rundownStarted = true;
+            _isShuttingDown = true;
+
+            // Wait for all shared and exclusive holders to release.
+            while (_activeCount > 0 || _exclusiveHeld)
+            {
+                _ = Monitor.Wait(_lock);
+            }
+
+            // Return a token; disposing it marks rundown complete and wakes other waiters.
+            return new RundownToken(this);
+        }
+    }
+
+    /// <summary>
+    /// Releases a shared acquisition and wakes waiters.
+    /// </summary>
+    public void ReleaseShared()
+    {
+        lock (_lock)
+        {
+            if (_activeCount <= 0)
+            {
+                throw new InvalidOperationException("No shared guards held");
+            }
+            _activeCount--;
+            // Wake up threads that might be waiting for capacity or for all shared to drain.
+            Monitor.PulseAll(_lock);
+        }
+    }
+
+    /// <summary>
+    /// Releases an exclusive acquisition and wakes waiters.
+    /// </summary>
+    public void ReleaseExclusive()
+    {
+        lock (_lock)
+        {
+            if (_exclusiveHeld == false)
+            {
+                throw new InvalidOperationException("No exclusive guard held");
+            }
+
+            _exclusiveHeld = false;
+            // Wake up threads waiting for exclusivity or rundown progress.
+            Monitor.PulseAll(_lock);
+        }
+    }
+
+    /// <summary>
+    /// Token returned by <see cref="AcquireSharedToken"/>; disposing it releases the shared hold.
+    /// </summary>
+    public sealed class ProtectionToken : IDisposable
+    {
+        private RundownGuard _owner;
+
+        internal ProtectionToken(RundownGuard owner)
+        {
+            _owner = owner;
+        }
+
+        public void Dispose()
+        {
+            _owner.ReleaseShared();
+        }
+    }
+
+    /// <summary>
+    /// Token returned by <see cref="AcquireExclusiveToken"/>; disposing it releases the exclusive hold.
+    /// </summary>
+    public sealed class ExclusiveToken : IDisposable
+    {
+        private RundownGuard _owner;
```

Comment:
NITPICK; make readonly
```suggestion
        private readonly RundownGuard _owner;
```